pkg/crypto/hash.go

```diff
@@ -0,0 +1,89 @@
+package crypto
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"strings"
+
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	// DefaultHasher is the default implementation for hashing things
+	// It outputs 32 Bytes and uses a SHA3-256 hash in the current configuration.
+	// Its generic security strength is 256 bits against preimage attacks,
+	// and 128 bits against collision attacks.
+	defaultHasher = basicHasher{sha3.New256()}
+)
+
+// Hash is a convenience function calling the default hasher
+func Hash(data ...interface{}) ([]byte, error) {
+	return defaultHasher.Hash(data...)
+}
+
+// HashToString is a convenience function calling the default hasher and encoding the result as hex string
+func HashToString(data ...interface{}) (string, error) {
+	hash, err := defaultHasher.Hash(data...)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(hash), nil
+}
+
+// Hasher provides a method for hashing arbitary data types
+type Hasher interface {
+	Hash(data ...interface{}) ([]byte, error)
+}
+
+type basicHasher struct {
+	hash hash.Hash
+}
+
+func (h basicHasher) Hash(args ...interface{}) ([]byte, error) {
+	h.hash.Reset()
+
+	for _, data := range args {
+		var (
+			reader io.Reader
+			wg     = &errgroup.Group{}
+		)
+
+		// setup reader for the data
+		switch d := data.(type) {
+		case io.Reader:
+			reader = d
+		case []byte:
+			reader = bytes.NewReader(d)
+		case string:
+			reader = strings.NewReader(d)
+		case fmt.Stringer:
+			reader = strings.NewReader(d.String())
+		default:
+			r, w := io.Pipe()
+			encoder := json.NewEncoder(w)
+			wg.Go(func() error {
+				defer w.Close()
+				return encoder.Encode(data)
+			})
+			reader = r
+		}
+
+		// hash all the data
+		_, err := io.Copy(h.hash, reader)
+		if err != nil {
+			return nil, err
+		}
+
+		// wait for the encoder
+		if err := wg.Wait(); err != nil {
```

Comment:
We don't need the waiting group and the wait anymore. You can just run:

```go
go func() error {
				defer w.Close()
				return encoder.Encode(data)
			}()
```

The copy will handle blocking and error handling for you. The only thing we need it to write in a separate go routine here. Also, it would one dependency from go-base (x/sync).

Reply:
Or we can leave waiting group and handle the error explicitly.

Reply:
Thats ignoring the marshaling error, but I get the gist. Give me a sec!

Reply:
updated.